In [2]:
import torch
from tinypytorch.data import get_local_data
from tinypytorch.model import initialize_parameters, Model
from tinypytorch.metrics import accuracy

In [3]:
#| hide
from tinypytorch.core import *

# tinypytorch

> A tiny version of pytorch from scratch

This file will become your README and also the index of your documentation.

## Install

```sh
pip install tinypytorch
```

### Train a neural network

#### Data

In [4]:
x_train, y_train, x_valid, y_valid = get_local_data()

In [5]:
x_train.shape, y_train.shape

(torch.Size([50000, 784]), torch.Size([50000]))

In [6]:
y_train[1]

tensor(0)

In [7]:
y_train[1:5]

tensor([0, 4, 1, 9])

#### Initialize hyperparameters

In [8]:
n, m = x_train.shape # num rows and columns

In [9]:
c = y_train.max() + 1

In [10]:
n, m, c

(50000, 784, tensor(10))

In [11]:
nh = 50 # num hidden

In [12]:
w1, b1, w2, b2 = initialize_parameters(m, nh)

In [13]:
w1.shape, b1.shape

(torch.Size([784, 50]), torch.Size([50]))

In [14]:
w2.shape, b2.shape

(torch.Size([50, 1]), torch.Size([1]))

- Training set's shape: (50000, 784)
- Weight's shape: (784, 50)
- Bias's shape: (50)

#### The first layer (Lin): (50000, 784) x (784, 50) + (50)

In [15]:
model = Model(w1, b1, w2, b2)

In [16]:
loss = model(x_train, y_train)

Model.__call__
l=<tinypytorch.model.Lin object at 0x7f85497baee0>
Lin.forward
inp=torch.Size([50000, 784])
w=torch.Size([784, 50])
b=torch.Size([50])
output.shape=torch.Size([50000, 50])
x.shape=torch.Size([50000, 50])
Model.__call__
l=<tinypytorch.model.ReLU object at 0x7f85497ba7c0>
x.shape=torch.Size([50000, 50])
Model.__call__
l=<tinypytorch.model.Lin object at 0x7f85497baf40>
Lin.forward
inp=torch.Size([50000, 50])
w=torch.Size([50, 1])
b=torch.Size([1])
output.shape=torch.Size([50000, 1])
x.shape=torch.Size([50000, 1])


In [17]:
loss

tensor(29.8016)

In [18]:
model.backward()

### Example 2

In [19]:
bs = 64

In [21]:
xb = x_train[0:64]